dev-tools/omdb/src/bin/omdb/db.rs

```diff
@@ -4315,36 +4316,90 @@ async fn cmd_db_inventory_physical_disks(
         model: String,
         serial: String,
         variant: String,
+        firmware: String,
+        next_firmware: String,
     }
 
-    use db::schema::inv_physical_disk::dsl;
-    let mut query = dsl::inv_physical_disk.into_boxed();
+    use db::schema::inv_nvme_disk_firmware::dsl as firmware_dsl;
+    use db::schema::inv_physical_disk::dsl as disk_dsl;
+
+    let mut query = disk_dsl::inv_physical_disk.into_boxed();
     query = query.limit(i64::from(u32::from(limit)));
 
     if let Some(collection_id) = args.collection_id {
         query = query.filter(
-            dsl::inv_collection_id.eq(collection_id.into_untyped_uuid()),
+            disk_dsl::inv_collection_id.eq(collection_id.into_untyped_uuid()),
         );
     }
 
     if let Some(sled_id) = args.sled_id {
-        query = query.filter(dsl::sled_id.eq(sled_id.into_untyped_uuid()));
+        query = query.filter(disk_dsl::sled_id.eq(sled_id.into_untyped_uuid()));
     }
 
     let disks = query
-        .select(InvPhysicalDisk::as_select())
+        .left_join(
+            firmware_dsl::inv_nvme_disk_firmware.on(
+                firmware_dsl::inv_collection_id
+                    .eq(disk_dsl::inv_collection_id)
+                    .and(firmware_dsl::sled_id.eq(disk_dsl::sled_id))
+                    .and(firmware_dsl::slot.eq(disk_dsl::slot)),
+            ),
+        )
+        .select((
+            InvPhysicalDisk::as_select(),
+            Option::<InvNvmeDiskFirmware>::as_select(),
+        ))
         .load_async(&**conn)
         .await
         .context("loading physical disks")?;
 
-    let rows = disks.into_iter().map(|disk| DiskRow {
-        inv_collection_id: disk.inv_collection_id.into_untyped_uuid(),
-        sled_id: disk.sled_id.into_untyped_uuid(),
-        slot: disk.slot,
-        vendor: disk.vendor,
-        model: disk.model.clone(),
-        serial: disk.serial.clone(),
-        variant: format!("{:?}", disk.variant),
+    let rows = disks.into_iter().map(|(disk, firmware)| {
+        let mut active_firmware = "UNKNOWN".to_string();
+        let mut next_firmware = String::new();
+
+        'firmware: {
```

Comment:
I chose to make this a labeled block because I thought it was easier to read than having a bunch of nested if-let statements.  Happy to change it back if people prefer that style over this.

Reply:
what do you think about a
```rust
impl InvNvmeDiskFirmware {
    pub fn current_version(&self) -> Option<&str> {
        self.slot_firmware_versions
            .get(self.active_slot.0 as usize - 1) 
            .map(|v| v.as_ref())
            .flatten()
            .map(|v| v.as_str())
    }

    pub fn next_version(&self) -> Option<&str> {
        self.next_active_slot
            .and_then(|slot| {
                self.slot_firmware_versions.get(slot.0 as usize - 1)
            })
            .map(|v| v.as_ref())
            .flatten()
            .map(|v| v.as_str())
    }    
}
```

and then changing this block to something more like
```rust
let (active_firmware, next_firmware) =
    if let Some(firmware) = firmware.as_ref() {
        (firmware.current_version(), firmware.next_version())
    } else {
        (None, None)
    };

...
firmware: active_firmware.unwrap_or("UNKNOWN").to_string(),
next_firmware: next_firmware.unwrap_or("").to_string(),
```
?

maybe-bonus: clearer to see what happens if something is super horked up with firmware info?

Reply:
Yes! You are absolutely right that these should be in the impl of `InvNvmeDiskFirmware`.

Reply:
Fixed in [072fcac](https://github.com/oxidecomputer/omicron/pull/6275/commits/072fcacabd29e67ca5abd06fd48a63db2a62c2fc)